worker/modelcache/worker.go

```diff
@@ -0,0 +1,246 @@
+// Copyright 2018 Canonical Ltd.
+// Licensed under the AGPLv3, see LICENCE file for details.
+
+package modelcache
+
+import (
+	"sync"
+
+	"github.com/juju/errors"
+	"github.com/kr/pretty"
+	"github.com/prometheus/client_golang/prometheus"
+	"gopkg.in/juju/worker.v1"
+	"gopkg.in/juju/worker.v1/catacomb"
+
+	"github.com/juju/juju/core/cache"
+	"github.com/juju/juju/core/life"
+	"github.com/juju/juju/core/status"
+	"github.com/juju/juju/state"
+	"github.com/juju/juju/state/multiwatcher"
+)
+
+// Config describes the necessary fields for NewWorker.
+type Config struct {
+	Logger               Logger
+	StatePool            *state.StatePool
+	PrometheusRegisterer prometheus.Registerer
+	Cleanup              func()
+	// Notify is used primarily for testing, and is passed through
+	// to the cache.Controller. It is called every time the controller
+	// processes an event.
+	Notify func(interface{})
+}
+
+// Validate ensures all the necessary values are specified
+func (c *Config) Validate() error {
+	if c.Logger == nil {
+		return errors.NotValidf("missing logger")
+	}
+	if c.StatePool == nil {
+		return errors.NotValidf("missing state pool")
+	}
+	if c.PrometheusRegisterer == nil {
+		return errors.NotValidf("missing prometheus registerer")
+	}
+	if c.Cleanup == nil {
+		return errors.NotValidf("missing cleanup func")
+	}
+	return nil
+}
+
+type cacheWorker struct {
+	config     Config
+	catacomb   catacomb.Catacomb
+	controller *cache.Controller
+	changes    chan interface{}
+	watcher    *state.Multiwatcher
+	mu         sync.Mutex
+}
+
+// NewWorker creates a new cacheWorker, and starts an
+// all model watcher.
+func NewWorker(config Config) (worker.Worker, error) {
+	if err := config.Validate(); err != nil {
+		return nil, errors.Trace(err)
+	}
+	w := &cacheWorker{
+		config:  config,
+		changes: make(chan interface{}),
+	}
+	controller, err := cache.NewController(
+		cache.ControllerConfig{
+			Changes: w.changes,
+			Notify:  config.Notify,
+		})
+	if err != nil {
+		return nil, errors.Trace(err)
+	}
+	w.controller = controller
+	if err := catacomb.Invoke(catacomb.Plan{
+		Site: &w.catacomb,
+		Work: w.loop,
+		Init: []worker.Worker{w.controller},
+	}); err != nil {
+		return nil, errors.Trace(err)
+	}
+	return w, nil
+}
+
+// Report returns information that is used in the dependency engine report.
+func (c *cacheWorker) Report() map[string]interface{} {
+	return c.controller.Report()
+}
+
+func (c *cacheWorker) loop() error {
+	defer c.config.Cleanup()
+	pool := c.config.StatePool
+
+	allWatcherStarts := prometheus.NewGauge(
```

Comment:
Why a gauge rather than a counter? This is a monotonically increasing value.

Reply:
OK, changed.